uktrade/data-hub-frontend
# Task component (Redux-Saga)

## Description of change

Added the redux-saga based `Task` component, an abstraction of an asynchronous task which can be in progress, possibly fail and be retried.

I couldn't write unit tests for this because it blows up on `import` statements.

## Test instructions

The component is not used anywhere yet, but the plan is to use it first in the _my companies list_ in the dashboard. To see it in action you can replace the contents of `src/client/index.jsx` with the following snippet, which is roughly the way the _my companies list_ will be done:

```jsx
// Paste me to src/client/index.jsx
import React from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import { createStore, combineReducers, applyMiddleware } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly'
import createSagaMiddleware from 'redux-saga'

import Task from './components/Task'
import tasksSagaFactory from './components/Task/saga'
import tasksReducer from './components/Task/reducer'
import { connect } from 'react-redux'

const companyListsReducer = (
  state = { listDetails: {} },
  { type, ...action }
) => {
  switch (type) {
    case 'COMPANY_LISTS__LISTS_LOADED':
      return {
        ...state,
        lists: action.result,
        selectedId: action.result[0].id,
      }
    case 'COMPANY_LISTS__COMPANIES_LOADED':
      return {
        ...state,
        listDetails: {
          ...state.listDetails,
          [action.payload]: action.result,
        },
      }
    case 'COMPANY_LIST__SELECT':
      return { ...state, selectedId: action.id }
    default:
      return state
  }
}

const dummyFetchCompanyistsTask = () =>
  new Promise((resolve) =>
    setTimeout(resolve, 1000, [
      {
        name: 'Foo',
        id: 'foo',
      },
      {
        name: 'Bar',
        id: 'bar',
      },
      {
        name: 'Baz',
        id: 'baz',
      },
    ])
  )

const dummyFetchSingleCompanyListTask = (id) =>
  new Promise((resolve, reject) =>
    id == 'bar'
      ? setTimeout(reject, 1000, Error('foo failed'))
      : setTimeout(
          resolve,
          1000,
          {
            foo: ["Ian's campervans", "Peter's RVs"],
            bar: ['Blah blah'],
            baz: ['Aaaa', 'Bbbbb', 'Ccccc'],
          }[id]
        )
  )

const CompanyLists = connect((state) => state.companyLists)(
  ({ lists, selectedId, listDetails, dispatch }) =>
    lists ? (
      <>
        Company lists
        <select
          onChange={(e) =>
            dispatch({
              type: 'COMPANY_LIST__SELECT',
              id: e.target.value,
            })
          }
        >
          {lists.map(({ id, name }) => (
            <option key={id} value={id}>
              {name}
            </option>
          ))}
        </select>
        {listDetails[selectedId] ? (
          <ul>
            {listDetails[selectedId].map((company) => (
              <li key={company} id={company}>
                {company}
              </li>
            ))}
          </ul>
        ) : (
          <Task.Status
            name="Company list"
            id={selectedId}
            progressMessage="Loading companies"
            startOnRender={{
              payload: selectedId,
              onSuccessDispatch: 'COMPANY_LISTS__COMPANIES_LOADED',
            }}
          />
        )}
      </>
    ) : (
      <Task.Status
        name="Company lists"
        id="dashboard"
        progressMessage="Loading company lists"
        startOnRender={{
          onSuccessDispatch: 'COMPANY_LISTS__LISTS_LOADED',
        }}
      />
    )
)

const sagaMiddleware = createSagaMiddleware()

const store = createStore(
  combineReducers({
    companyLists: companyListsReducer,
    tasks: tasksReducer,
  }),
  composeWithDevTools(applyMiddleware(sagaMiddleware))
)

sagaMiddleware.run(
  tasksSagaFactory({
    // Register tasks
    'Company lists': dummyFetchCompanyistsTask,
    'Company list': dummyFetchSingleCompanyListTask,
  })
)

ReactDOM.render(
  <Provider store={store}>
    <CompanyLists />
  </Provider>,
  document.getElementById('react-app'),
)

```

## Checklist

[//]: # "When submitting a PR make sure the code review guidelines have been satisfied.
https://github.com/uktrade/data-hub-frontend/blob/master/docs/Code%20review%20guidelines.md"
